src/libsystemd-network/sd-dhcp-server.c

```diff
@@ -1214,3 +1280,42 @@ int sd_dhcp_server_set_callback(sd_dhcp_server *server, sd_dhcp_server_callback_
 
         return 0;
 }
+
+int sd_dhcp_server_add_static_lease(sd_dhcp_server *server, sd_dhcp_static_lease *v) {
+        _cleanup_(sd_dhcp_static_lease_unrefp) sd_dhcp_static_lease *old = NULL;
+        int r;
+
+        assert_return(server, -EINVAL);
+        assert_return(v, -EINVAL);
+
+        r = hashmap_ensure_allocated(&server->static_leases_by_client_id, &dhcp_static_lease_hash_ops);
+        if (r < 0)
+                return r;
+
+        old = hashmap_get(server->static_leases_by_client_id, &v->client_id);
+        r = hashmap_replace(server->static_leases_by_client_id, &v->client_id, v);
+        if (r < 0) {
+                hashmap_remove(server->static_leases_by_client_id, &v->client_id);
+                return r;
```

Comment:
we generally strive for functions that fail fully or not at all. i.e. that when the fail they do not alter the object at all and never leave "half-failed" operations around. But here, if things fail you drop the old entry, but i don't understand why